File: Search in Rotated Sorted Array/search_in_rotated_sorted_array.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int left = 0;
        int right = nums.size() - 1;

        while (left <= right) {
            int middle = left + (right - left) / 2;
            if (nums[middle] == target) {
                return middle;
            } else if (nums[left] <= nums[middle]) { // we are in the left
                if (target > nums[middle] || target < nums[left]) {
                    // search right
                    left = middle + 1;
                } else {
                    right = middle - 1;
                }
            } else { // we are in the right
                if (target > nums[right] || nums[middle] > target) {
                    // search left
                    right = middle - 1;
                } else {
                    left = middle + 1;
                }
            }
        }
        return -1;
    }
};

File: 33-search-in-rotated-sorted-array/search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l, r = 0, len(nums) - 1 

        while l <= r:
            mid = l + (r - l) // 2
            
            if nums[mid] == target:
                return mid
            
            # CASE 1: The left half is sorted
            if nums[l] <= nums[mid]:
                # Is the target inside this sorted left half?
                if nums[l] <= target < nums[mid]:
                    r = mid - 1 # Search left
                else:
                    l = mid + 1 # Search right
            
            # CASE 2: The right half must be sorted
            else:
                # Is the target inside this sorted right half?
                if nums[mid] < target <= nums[r]:
                    l = mid + 1 # Search right
                else:
                    r = mid - 1 # Search left
                    
        return -1